packages/batch-submitter/src/batch-submitter.ts

```diff
@@ -0,0 +1,234 @@
+    const totalElementsToAppend = blocks.length
+
+    // Generate contexts
+    const contexts: BatchContext[] = []
+    let lastBlockIsSequencerTx = false
+    const groupedBlocks: Array<{
+      sequenced: BatchElement[]
+      queued: BatchElement[]
+    }> = []
+    for (const block of blocks) {
+      if (
+        (lastBlockIsSequencerTx === false && block.isSequencerTx === true) ||
+        groupedBlocks.length === 0
+      ) {
+        groupedBlocks.push({
+          sequenced: [],
+          queued: [],
+        })
+      }
+      const cur = groupedBlocks.length - 1
+      block.isSequencerTx
+        ? groupedBlocks[cur].sequenced.push(block)
+        : groupedBlocks[cur].queued.push(block)
+      lastBlockIsSequencerTx = block.isSequencerTx
+    }
+    for (const groupedBlock of groupedBlocks) {
+      contexts.push({
+        numSequencedTransactions: groupedBlock.sequenced.length,
+        numSubsequentQueueTransactions: groupedBlock.queued.length,
+        timestamp:
+          groupedBlock.sequenced.length > 0
+            ? groupedBlock.sequenced[0].timestamp
+            : 0,
+        blockNumber:
+          groupedBlock.sequenced.length > 0
+            ? groupedBlock.sequenced[0].blockNumber
+            : 0,
+      })
+    }
+
+    // Generate sequencer transactions
+    const transactions: string[] = []
+    for (const block of blocks) {
+      if (!block.isSequencerTx) {
+        continue
+      }
+      let encoding: string
+      if (block.sequencerTxType === TxType.EIP155) {
+        encoding = ctcCoder.eip155TxData.encode(block.txData as EIP155TxData)
+      } else if (block.sequencerTxType === TxType.EthSign) {
+        encoding = ctcCoder.ethSignTxData.encode(block.txData as EthSignTxData)
+      } else if (block.sequencerTxType === TxType.createEOA) {
+        encoding = ctcCoder.createEOATxData.encode(
+          block.txData as CreateEOATxData
+        )
+      }
+      transactions.push(encoding)
+    }
+
+    return {
+      shouldStartAtBatch: shouldStartAtIndex - 1,
+      totalElementsToAppend,
+      contexts,
+      transactions,
+    }
+  }
+
+  public async _getL2BatchElement(blockNumber: number): Promise<BatchElement> {
+    const block = (await this.l2Provider.getBlockWithTransactions(
+      blockNumber
+    )) as L2Block
+    const txType = block.transactions[0].meta.txType
+
+    if (this._isSequencerTx(block)) {
+      if (txType === TxType.EIP155 || txType === TxType.EthSign) {
+        return this._getDefaultEcdsaTxBatchElement(block)
+      } else if (txType === TxType.createEOA) {
+        return this._getCreateEoaBatchElement(block)
+      } else {
+        throw new Error('Unsupported Tx Type!')
+      }
+    } else {
+      return {
+        stateRoot: block.stateRoot,
+        isSequencerTx: false,
+        sequencerTxType: undefined,
+        txData: undefined,
+        timestamp: block.timestamp,
+        blockNumber: block.transactions[0].meta.l1BlockNumber,
```

Comment:
This is kinda unsafe and i see this function being called without `try/catch`

Reply:
I wrapped the call to `submitNextBatch()` in a try catch which I think should cover all the errors because this is entirely stateless I was thinking it would be easier to catch everything there & just retry the whole sequence